src/wrapper/objects/weak_global_ref.rs

```diff
@@ -0,0 +1,97 @@
+use std::{convert::From, sync::Arc};
+
+use log::{debug, warn};
+
+use crate::{errors::Result, objects::JWeak, sys, JNIEnv, JavaVM};
+
+/// A weak global JVM reference. **It doesn't protect underlaying object from been
+/// garbage collected.** Still, `WeakGlobalRef` is allowed to outlive the `JNIEnv` that
+/// it came from and can be used in other threads.
+///
+/// `WeakGloablRef` may be created via
+/// [JNIEnv::new_weak_global_ref](../struct.JNIEnv.html#method.new_weak_global_ref).
+///
+/// `WeakGloablRef` doesn't allow access to the underlying object but may be
+/// upgraded into _potentially null_ [GlobalRef](struct.GlobalRef.html) via
```

Comment:
I don't believe it's a good idea to use JavaVM because the thread may not be attached to the JavaVM. 

We're forcing thread attaching for Drop because we don't have any other option. For all other purposes it's User's responsibility to provide correct JNIEnv.

I have no objections as per creation the methods for WeakGlobalRef instead of JNIEnv. But the methods should require JNIEnv to be provided as a parameter.

What do you think?

Reply:
For locals, having the current thread attached is the only way to do it; for globals that is a sensible way, because thread attachment is expensive. 

My thought was to get the env of an already attached thread via JavaVM#get_env, and failing in case it is not.

I think the benefit of passing JNIEnv is that it is more explicit that it is needed; the benefit of taking it under the hood is that the methods require less arguments. But there is a cost with the 2nd approach: there are more failure modes, and it can be used incorrectly (without JNIEnv). I lean towards passing an argument.